video/out/vo_sixel.c

```diff
@@ -0,0 +1,429 @@
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+#include <libavutil/rational.h>
+#include <libavutil/time.h>
+#include <libswscale/swscale.h>
+#include <sixel.h>
+#include <termios.h>
+
+#include "config.h"
+#include "options/m_config.h"
+#include "osdep/terminal.h"
+#include "sub/osd.h"
+#include "vo.h"
+#include "video/sws_utils.h"
+#include "video/mp_image.h"
+
+#define IMGFMT IMGFMT_RGB24
+
+#define ESC_HIDE_CURSOR "\033[?25l"
+#define ESC_RESTORE_CURSOR "\033[?25h"
+#define ESC_CLEAR_SCREEN "\033[2J"
+#define ESC_GOTOXY "\033[%d;%df"
+#define ESC_TERMINATE_DCS "\033\\"
+#define ESC_SAVE_XY_CURSOR_LOCATION "\033[%d;%dH\0337"
+#define ESC_RESTORE_SAVED_CURSOR "\0338"
+#define ESC_USE_GLOBAL_COLOR_REG "\033[?1070l"
+
+
+struct vo_sixel_opts {
+    int diffuse;
+    int width;
+    int height;
+    int reqcolors;
+    int fixedpal;
+    int threshold;
+    int top;
+    int left;
+};
+
+#define OPT_BASE_STRUCT struct vo_sixel_opts
+static const struct m_sub_options vo_sixel_conf = {
+    .opts = (const m_option_t[]) {
+        {"vo-sixel-diffusion", OPT_CHOICE(diffuse,
+            {"auto", DIFFUSE_AUTO},
+            {"none", DIFFUSE_NONE},
+            {"atkinson", DIFFUSE_ATKINSON},
+            {"fs", DIFFUSE_FS},
+            {"jajuni", DIFFUSE_JAJUNI},
+            {"stucki", DIFFUSE_STUCKI},
+            {"burkes", DIFFUSE_BURKES},
+            {"arithmetic", DIFFUSE_A_DITHER},
+            {"xor", DIFFUSE_X_DITHER})},
+        {"vo-sixel-width", OPT_INT(width)},
+        {"vo-sixel-height", OPT_INT(height)},
+        {"vo-sixel-reqcolors", OPT_INT(reqcolors)},
+        {"vo-sixel-fixedpalette", OPT_INT(fixedpal)},
+        {"vo-sixel-color-threshold", OPT_INT(threshold)},
+        {"vo-sixel-offset-top", OPT_INT(top)},
+        {"vo-sixel-offset-left", OPT_INT(left)},
+        {0}
+    },
+    .defaults = &(const struct vo_sixel_opts) {
+        .diffuse = DIFFUSE_ATKINSON,
+        .width = 320,
+        .height = 240,
+        .reqcolors = 256,
+        .fixedpal = 0,
+        .threshold = 0,
+        .top = 1,
+        .left = 1,
+    },
+    .size = sizeof(struct vo_sixel_opts),
+};
+
+struct priv {
+    struct vo_sixel_opts *opts;
+
+    sixel_output_t *output;
+    sixel_dither_t *dither;
+    sixel_dither_t *testdither;
+    uint8_t        *buffer;
+
+    int image_height;
+    int image_width;
+    int image_format;
+
+    struct mp_image *frame;
+    struct mp_sws_context *sws;
+};
+
+static const unsigned int depth = 3;
+
+static void save_image_offset_location(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+    int top = priv->opts->top;
+    int left = priv->opts->left;
+    int terminal_width = 0;
+    int terminal_height = 0;
+
+    terminal_get_size(&terminal_width, &terminal_height);
+
+    // Make sure that the user specified top offset
+    // lies in the range 1 to TERMINAL_HEIGHT
+    // Otherwise default to the topmost row
```

Comment:
I don't understand this. The allowed offset (horizontal and vertical) should be such that it both starts and ENDs within the terminal area. Why do you test only where it starts?

Reply:
Because you can start with an initial terminal that is smaller in dimensions, and start playback with only like some parts of video being displayed, and later resize the terminal and have sixel show the other parts to the right and bottom.

Reply:
How would it work if parts of the video end outside the terminal? Does it still work smoothly? no extra scrolling or flickering?

Reply:
If the height exceeds then there is flickering. But if the width exceeds, there is no issue as far as I have observed, and on increasing the window it smoothly displays.